doc/jira_bridge.md

```diff
@@ -0,0 +1,365 @@
+will query the list of available transitions for the issue. If a transition
+exists to the desired state the bridge will attempt to execute the transition.
+It does not currently support assigning any fields during the transition so if
+any fields are required the transition will fail during export and the status
+will be out of sync.
+
+### JIRA Changelog
+
+Some operations on JIRA issues are visible in a timeline view known as the
+`changelog`. The JIRA cloud product provides an
+`/issue/{issueIdOrKey}/changelog` endpoint which provides a paginated view but
+the JIRA server product does not. The changelog is visible by querying the issue
+with the `expand=changelog` query parameter. Unfortunately in this case the
+entire changelog is provided without paging.
+
+Each changelog entry is identified with a unique string `id`, but within a
+single changelog entry is a list of multilple fields that are modified. In other
+words a single "event" might atomically change multiple fields. As an example,
+when an issue is closed the `"status"` might change to `"closed"` and the
+`"resolution"` might change to `"fixed'`.
+
+When a changelog entry is imported by the JIRA bridge, each individual field
+that was changed is treated as a separate `git-bug` operation. In other words a
+single JIRA change event might create more than one `git-bug` operation.
+
+However, when a `git-bug` operation is exported to JIRA it will only create a
+single changelog entry. Furthermore, when we modify JIRA issues over the REST
+API JIRA does not provide any information to associate that modification event
+with the changelog. We must, therefore, herustically match changelog entries
+against operations that we performed in order to not import them as duplicate
+events. In order to assist in this matching proceess, the bridge will record the
+JIRA server time of the response to the `POST` (as reported by the `"Date"`
+response header). During import, we keep an iterator to the list of `git-bug`
+operations for the bug mapped to the Jira issue. As we walk the JIRA changelog,
+we keep the iterator pointing to the first operation with an annotation which is
+*not before* that changelog entry. If the changelog entry is the result of an
+exported `git-bug` operation, then this must be that operation. We then scan
+through the list of changeitems (changed fields) in the changelog entry, and if
+we can match a changed field to the candidate `git-bug` operation then we have
+identified the match.
+
+### Unlogged Changes
+
+Comments (creation and edition) do not show up in the JIRA changelog. However
+JIRA reports both a `created` and `updated` date for each comment. If we
+import a comment which has an `updated` and `created` field which do not match,
+then we treat that as a new comment edition. If we do not already have the
+comment imported, then we import an empty comment followed by a comment edition.
```

Comment:
Operation order in the data-structure is meaningful and unless you manually alter it it's not possible to add an operation *before* an existing one. You wont be able to add a revision before another one.

Reply:
I think you are trying to solve the wrong problem here. The bridge should not import a comment that it already has. Maybe you can compute some sort of signature for your comment, maybe `hash(comment || date)` or something ?

Reply:
> Operation order in the data-structure is meaningful and unless you manually alter it it's not possible to add an operation before an existing one. You wont be able to add a revision before another one.

Wait, really? Maybe I misunderstood [this part][1]. I thought the whole idea of maintaining a list of edit operations was so that one could insert an interleaved operation. If not, that is fine. I guess it wont make much difference then whether we create the initial comment as empty or with the current text. I think it's important to actually record the edit operation, however, so that we know we don't have to do it on the next import of this issue. 

> The bridge should not import a comment that it already has.

It doesn't import a comment it already has. Sorry I wasn't clear: If the bridge already has a comment, but it observes a new revision, it does not import a whole new comment, it edits the existing comment.

[1]: https://github.com/MichaelMure/git-bug/blob/master/doc/model.md#you-cant-rely-on-the-time-provided-by-other-people-their-clock-might-by-off-for-anything-other-than-just-display

Reply:
Lamport clock are used to order bugs with each other (creation and last edition), but within a bug the order of the operation are what matter. You can see it as a sort of blockchain where the more operation you add, the more you prevent tempering the history. That said, I'm considering at the moment to relax this purely linear chain constraint and have a DAG of Operation to get a more robust conflict merge (see https://github.com/MichaelMure/git-bug/issues/212 but it's quite messy). If that happen, Lamport clock or similar might have a bigger role to play.

In any case, bridges should consider the data structure as append only.

> If the bridge already has a comment, but it observes a new revision, it does not import a whole new comment, it edits the existing comment.

My comment still stand: if a comment edition is generated for each import while no change happened on the Jira side, it's going to be problematic.

Reply:
> My comment still stand: if a comment edition is generated for each import while no change happened on the Jira side, it's going to be problematic.

This is not the case. The logic is:

>  If the bridge already has a comment, but it observes a **new revision**, it does not import a whole new comment, it edits the existing comment.

The logic should not create a new comment edition for each import. It should create a new comment edition only if it sees an previously un-imported JIRA "revision" in the changelog... And in testing I believe this is accurately implemented.

Reply:
perfect then :+1: